### PR TITLE
feat(core): AFTER_STAGE stages can now be planned post task completion

### PR DESCRIPTION
**Pull request by ajordens**
This allows for interesting scenarios where a stage could have a `Task`
responsible for building the context that is subsequently used to build
a dynamic set of synthetic stages.

Currently this logic must reside entirely in `aroundStages()` and you
lose the ability to arbitrarily retry with backoff's, etc. that normal
`Task` execution provides.

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/cluster/RollbackClusterStage.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.netflix.spinnaker.orca.pipeline.StageDefinitionBuilder.newStage;
+
+@Component
+public class RollbackClusterStage implements StageDefinitionBuilder {
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/cluster/RollbackClusterStage.java]: Consider this a proof of concept that takes advantage of the new `afterStages()` support.

The implementation of `DetermineRollbackCandidatesTask` is not entirely complete.

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandlerTest.kt [hunk 2]
```diff
@@ -42,9 +43,10 @@ object CompleteStageHandlerTest : SubjectSpek<CompleteStageHandler>({
   val publisher: ApplicationEventPublisher = mock()
   val clock = fixedClock()
   val contextParameterProcessor: ContextParameterProcessor = mock()
+  val stageDefinitionBuilderFactory: StageDefinitionBuilderFactory = mock()
 
   subject(GROUP) {
-    CompleteStageHandler(queue, repository, publisher, clock, contextParameterProcessor)
+    CompleteStageHandler(queue, repository, publisher, clock, contextParameterProcessor, stageDefinitionBuilderFactory)
```
Comment (ajordens) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandlerTest.kt]: If we're happy with the approach, I'll add an additional test in here to validate that `planAfterStages()` is invoked as expected.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/StageDefinitionBuilder.java [hunk 3]
```diff
@@ -50,6 +50,10 @@ default void taskGraph(
     return emptyList();
   }
 
+  default @Nonnull List<Stage> afterStages(@Nonnull Stage stage) {
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/StageDefinitionBuilder.java]: My expectation is that we can subsequently start moving `STAGE_AFTER` definitions from `aroundStages()` to `afterStages()`.

Comment (robfletcher) [reply orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/StageDefinitionBuilder.java]: What happens if you have after stages in both methods?

Comment (ajordens) [reply orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/StageDefinitionBuilder.java]: Your `afterStages()` would not get planned ... the `CompleteStageHandler` first calls `Stage.firstAfterStages()` to verify that there aren't any synthetic after stages to begin with.

**Pull request merged**